### PR TITLE
use sourcemaps to get a more accurate bundle-report

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

Closes https://github.com/parcel-bundler/parcel/issues/4368

This Pull Request is a work in progress to use sourcemaps to get a more accurate build report after minification and treeshaking.

It is not entirely accurate as our sourcemaps aren't entirely accurate or complete, but it's a lot better than before.

Before:
Screenshot 2020-04-11 at 13 19 32

After:
Screenshot 2020-04-11 at 13 18 47

### CODE REVIEW

File: packages/reporters/cli/src/bundleReport.js [hunk 1]
```diff
@@ -16,9 +17,14 @@ const COLUMNS = [
...
 
-export default function bundleReport(bundleGraph: BundleGraph) {
+export default async function bundleReport(
+  bundleGraph: BundleGraph,
+  fs: FileSystem,
+  projectRoot: FilePath,
+) {
   // Get a list of bundles sorted by size
-  let {bundles} = generateBundleReport(bundleGraph);
+  // $FlowFixMe not entirely sure why this fails?
```
Comment (mischnic) [inline packages/reporters/cli/src/bundleReport.js:26]: You access `bundle.totalAssets`  but `totalAssets` does't exist in `BundleStats`.

**Comment by devongovett** [comment]
Nice! Have you seen https://github.com/danvk/source-map-explorer? I believe the underlying library can be used standalone as well.

**Comment by DeMoorJasper** [comment]
@devongovett wow I didn't know that already existed. Looks pretty cool, although I don't think we should include a large library for what's essentially just a for loop over the mappings.

**Comment by wbinnssmith** [comment]
In the "after" screenshot, kitchen-sink.[hash].js has a couple 0 byte assets — is that accurate?

**Comment by DeMoorJasper** [comment]
@wbinnssmith it should be relatively accurate, it probably means there's no mappings with the origin to that asset.

**Comment by DeMoorJasper** [comment]
@wbinnssmith just double checked and it's a bug with the sourcemap. When throwing it in the sourcemap visualiser it shows up as `not found`. But the size is incorrect and should be correct once the sourcemap is correct.

Screenshot 2020-04-14 at 21 06 16

File: packages/core/utils/src/generateBuildMetrics.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+      let c = bundleContents[i];
+      if (
+        currMapping.generated.line === currLine &&
+        currMapping.generated.column <= currColumn &&
+        (nextMapping.generated.line > currLine ||
+          (nextMapping.generated.line === currLine &&
+            nextMapping.generated.column > currColumn))
+      ) {
+        sourceContents[currMapping.source] += c;
```
[the remark this answers is not in the document]
Comment (DeMoorJasper) [reply packages/core/utils/src/generateBuildMetrics.js:75]: That would work as well

File: packages/core/utils/src/generateBuildMetrics.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        index[a.filePath] = true;
+        return true;
+      } else {
+        return false;
+      }
+    })
+    .map(asset => {
+      let foundSize =
+        sourcemapSizes &&
+        sourcemapSizes.find(s => s.filePath === asset.filePath);
```
[the remark this answers is not in the document]
Comment (DeMoorJasper) [reply packages/core/utils/src/generateBuildMetrics.js:121]: You're right that would be better

**Comment by DeMoorJasper** [comment]
@devongovett I'm not sure about any performance cost, I doubt it would be a very large impact, it's just a relatively simple loop. Will check with the threejs benchmark

**Comment by mischnic** [comment]
> is this a known issue?

Not really, but I was aware of how the threejs benchmark is failing. Investigating now...

File: packages/core/utils/package.json [hunk 4]
```diff
@@ -21,6 +21,7 @@
     "@parcel/diagnostic": "^2.0.0-alpha.3.1",
     "@parcel/logger": "^2.0.0-alpha.3.1",
     "@parcel/markdown-ansi": "^2.0.0-alpha.3.1",
+    "@parcel/source-map": "^2.0.0-alpha.4.9",
```
Comment (mischnic) [inline packages/core/utils/package.json:24]:
```suggestion
    "@parcel/source-map": "2.0.0-alpha.4.9",
```

**Comment by mischnic** [comment]
If you replace `export {copy1};` with `console.log(copy1);` as a workaround, the error should go away.

**Comment by DeMoorJasper** [comment]
@devongovett I ran the threejs benchmark and it takes about 1-2 seconds on my machine to show the build report, so maybe it's a good idea to add the flag?

**Comment by wbinnssmith** [comment]
What should we do if someone wants scope hoisting but doesn't want source maps? Should we generate sourcemaps so we can create a meaningful report, but never write them as output?

cc @mischnic re: #4528

**Comment by mischnic** [comment]
> What should we do if someone wants scope hoisting but doesn't want source maps? Should we generate sourcemaps so we can create a meaningful report, but never write them as output?

I thought about that as well. I guess the question is if disabling sourcemaps completely (e.g. also for babel) has a significant performance benefit.

**Comment by DeMoorJasper** [comment]
@wbinnssmith It depends on how much anyone cares about performance. I think it saves us quite some time to not generate any sourcemaps. Maybe it should be part of the `--detailed-report` flag Devon mentioned. In case that flag is defined always generate source-map and the detailed report.

This report should still work without a sourcemap, but it would be identical to before this PR...

**Pull request merged**